warm diggety: a first crack at geometric algebra ('hestenic.cljc')

vanderoops opened the pull request:
Here's another geometric algebra note for when we get this ported over: https://marctenbosch.com/quaternions/

Commit:
warm diggety: a first crack at geometric algebra ('hestenic.cljc')

Commit:
how about a little equality, scarecrow?

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+
+
+;;
+;; A Gradeling is a bevy of Bladoids of the same grade. It may or may not be
```
what's a bevy of bladoids?

vanderoops replied:
(as discussed yesterday)

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+(defn gradeling [grade-or-bladoids]
+  (if
+      (integer? grade-or-bladoids)
+    (Gradeling. grade-or-bladoids ())
```
I think you depend on these being vectors in a few spots, so prefer `[]`.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  IGradeling
+  (bladoids [_] bldds)
+  (bladoid-with-basis [this b]
+    (first (filter (fn [el] (= (basis el) b))
```
I wouldn't do it here but just for Clojure fun you could also:

```clojure
(some #(when (= b (basis %)) %)
      (bladoids this))
```

vanderoops replied:
any reason this'd be preferred? `some`'s vague and indeterminate connotation (in english) feels like a mmiscue here.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  Object
+  (toString [_] (str "g" gr "{"
+                     (reduce (fn [sofar bl]
+                               (str sofar (if (empty? sofar) "" " ") bl))
```
`(clojure.string/join " " bldds)` for the guts.

vanderoops replied:
done.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+
+  Object
+  (toString [_] (str "MV["
+                     (reduce (fn [sofar gr]
```
`(clojure.string/join " " grdlgs)` for the guts

vanderoops replied:
prima! toll! fabelhaft! (schon gemacht)

vanderoops replied:
digging it everywhere all the time!

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+(defprotocol IVect
+  (coefs [this]))
+
+(deftype Vect [cfs]
```
Add metadata too!

vanderoops replied:
(have questions for you re: the metadata opportunity here in deftype land...)

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+;; of the whole algebra, y' see-- it's worth representing them explicitly.
+;;
+
+(defprotocol IVect
```
same note as above on protocols vs a bare function with a type check.

vanderoops replied:
aye! to be wrassled to the mat...

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  Object
+  (toString [_] (str "v" cfs)))
+
+(defn vect [& cs] (Vect. cs))
```
you can add

```clojure
(:refer-clojure :exclude [vector])
```

to your namespace declaration and call this `vector` explicitly.

vanderoops replied:
i am making a sizzling-sharp decision not to conflate the idea of a programming language vector with a geometric one; the names must therefore remain unambiguously distinct, without regard to namespace or any other implicit context. for this purpose, i've been lucky that `Vect` is largely untrampled in C++, Smalltalk, etc., and have worked a tattered flag into the thin poor soil accordingly...

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+        (and (number? otha) (zero? otha))))
+  (grade-part [this n] this))
+
+(defn zero-element []
```
How about doing

```clojure
(def zero-element (ZeroElement.))
```

So you just make this once?

vanderoops replied:
could do.

yeah, that's nice.

done!

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (neg [this] this)
+  (dif [this otha] (neg otha))
+  (prd [this otha] this)
+  (inv [this] (/ 1 0))
```
this is cute, but I'd recommend an explicit exception!

vanderoops replied:
cool. that properly requires a bit of thought and design.

well, actually, on second breath i'm not sure -- one organism's _cute_ is another's can you think of a more compact depiction of noninvertibility?

anyway, let's jaw!

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (prd [this otha] this)
+  (inv [this] (/ 1 0))
+  (quo [this otha]
+    (let [invvy (inv otha)]
```
I'd recommend an explicit exception if that thing is `zero?` (test it explicitly), otherwise return. Skip the work of actually inverting the other side. I know you're doing this to possibly trigger the exception...

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (wdg [this otha] this)
+  (eq? [this otha]
+    (or (instance? (class this) otha)
+        (and (number? otha) (zero? otha))))
```
use sicmutils.value to start getting us ready for cljs support: `v/number?`, `v/zero?`

vanderoops replied:
again-again: we need to decide on a posture re: using stuff from sicmutils in the context of wanting hestenic to be standalone.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+;; a shadowy cabal of utility-interior mathy bits.
+;;
+
+(defn- bladoid-mult-bladoid [bll blr]
```
you can call this ` bladoid-*-bladoid` if you want, just noting!

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+    (Bladoid. (* (coef bll) (coef blr) (order-swap-parity basl basr))
+              (basis-prod basl basr))))
+
+(defn- mv-absorb-gradeling [emvy gr]
```
these short names do make it hard to read... I'm not sure what the abbreviation even means! I get the `mv` part, by `e`, `y`?

vanderoops replied:
it's phonetic!

in cooperation with the comment foregoing, all arguments (and internal `let`-variables) now subscribe to a system that afixes `l-` or `r-` or `acc-` onto the left side of `doid`, `grdl`, or `emvy` according as the argument is the left or right Bladoid-or-Gradeling-or-MV in an addition or multiplicatoin.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+;; a shadowy cabal of utility-interior mathy bits.
+;;
+
+(defn- bladoid-mult-bladoid [bll blr]
```
prefer `l-blade`, `r-blade`, yeah? easier to read... or is that just me?

vanderoops replied:
not only have i made that change here; the `l-arg` / `r-arg` thing is now consistently propagated everywhere! america loves it!

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (reduce mv-absorb-gradeling
+          mva (gradelings mvb)))
+
+(defn- bladoid-mult-gradeling [bl gr]
```
you see where the generic multimethods would be helpful :)

vanderoops replied:
in this case i fear that i do not! all these functions are 'interior' worker-helper-laborer routines that won't be exposed. it is, rather, the lot of `sum` and `prd` and `dif` (etc.) that will get wired up into the generic multimethods, no?

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (reduce (fn [emvy bldlg] (mv-absorb-mv
+                            emvy
+                            (bladoid-mult-gradeling bldlg grr)))
+          (MV. '())
```
prefer `[]`... or `{}` if you take my dictionary suggestion!

sritchie replied:
Also I think you discovered that `()` doesn't need a quote...

vanderoops replied:
i'm all about leaving behind the lists that clj looks down on in favor of []!

all the species here are now converted thusly.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+          (bladoids grl)))
+
+(defn- mv-mult-gradeling [emvy rt_grdl]
+  (reduce (fn [out_mv lt_grdl]
```
woah, first time I've seen underscores!

vanderoops replied:
still developing what will converge as my typographic schema for this language. suspect underscores will appear somewhere to denote something... but no longer here.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+
+(defn- mv-map-biexploded-gradelings-and-sum [funq mvl mvr]
+  (reduce mv-absorb-mv
+          (MV. ())
```
You might consider creating and binding this empty multivector once.

vanderoops replied:
sarantastic! done.

hm. looking at it in a bunch of code revealed that `empty-mv` wasn't as visually distinguished as `(MV. [])`... but problem solved by upgrading the name to `the-empty-mv`, which also de-ambiguifies empty's part of speech. have gone back and done `zero-element` --> `the-zero-element` as well.

and also laid in a fresh monosupply of `the-empty-gradeling` for good measure, and because the mood demands it.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+;; rung conversions: one, yea, unto another.
+;;
+
+(extend-type Bladoid
```
You can do these inside the type definition too, by the way.

vanderoops replied:
aye. though not all protocol implementations can be so done, because of forward reference issues.

this one'd work, i think, but i prefer it in a separate corral for cognitive-clustery goodness.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (asGradeling [this]
+    (Gradeling. 1 (filter (fn [bl] (not= 0 (coef bl)))
+                          (map-indexed
+                           (fn [ind cf] (bladoid cf (vector ind)))
```
Prefer the literal `[ind]` to `(vector ind)`, especially if you change `vect` to `vector` above!

vanderoops replied:
as `vect` ain't going nowheauh...

... can ye kindly explain the preference?

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+                    spankOutNothingness
+                    (gradelings this))
+          spnkd (filter
+                 (fn [grdl] (not (number? grdl)))
```
`v/number?`

vanderoops replied:
ibid..dibi

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (quo [this otha]
+    (prd this (inv otha)))
+  (dot [this otha]
+    (reduce +
```
no need for the `0`, `reduce` calls the 0-arity of its aggregator function if there is no initializer, and that will return `0`.

vanderoops replied:
sho' 'nuf.

vanderoops replied:
right. i kind of liked it originally for its signalling that it's specifically a scalar getting accumulated... but have now taken it out.

sritchie reviewed (commented):
Okay, so much good stuff here!!!

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (grade [_] gr)
+
+  Object
+  (toString [_] (str "g" gr "{"
```
Whoops, ignore the `print-dup` thing, you only need `print-method`.

sritchie commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+          (nil? neck)
+          (list head)
+          (same-basis? head neck)
+          (recur (sum head neck) (second trunk) (rest trunk))
```
If you want to start getting into transducer magic, this is faster....

```clojure
(defn- collapse [blades]
  (sequence
   (comp (partition-by basis)
         (map (partial reduce sum)))
   blades))
```

vanderoops replied:
splend.

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+      (integer? grade-or-bladoids)
+    (Gradeling. grade-or-bladoids ())
+    (let [gra (if (empty? grade-or-bladoids)
+                -1
```
the intent of this signal (-1, here) is that the Gradeling exists but does not yet contain any Bladoids; the first one 'added' will define the grade foreverafter. thus doesn't feel like a job for a singleton (though let's maybe discuss?).

a reason to have as the sentinel value (which could of course be anything other than a non-negative integer)  -1 is that we use this approach in non-lisp implementations; and in languages for which the type of that 'field' is already defined to be an integer, we don't have the possibility of making such a sentinel be e.g. a symbol...

which implicitly asserts value in some pseudo-isomorphism of implementations across different languages...

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+(defn- order-swap-count [basl basr]
+  (loop [cnt 0
+         lind (dec (count basl))]
+    (if (< lind 0)
```
for me, those predicates (`pos?` et al.) pertain meaningfully to interrogating the essence of an immutable thing; whereas here we're doing what amounts to a low-level imperative-programming loop, with temporary index variables that are changing and whose instantaneous state needs to be tested. the `(< x 0)` style feels consonant with that. a matter of taste, of course...

MagusMachinae replied:
While predicates mostly seem to be implemented in Java for speed, taking `pos?` as a specific example, those implementations are still doing an inequality check at the end of the day, since `pos?` is a wrapper around a function called `isPos`. No prizes for guessing how that's implemented.  A rose by any other name...

The important idea here is binding that procedure to a name that scans faster, clearly communicates the aglorithm being used, and has less arguments you have to worry about getting wrong. I think it only feels consonant because the expectation that looping conditions look like that has been burnt into our minds by decades of C-overexposure.

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+      cnt
+      (let [lval (nth basl lind)
+            sc (scoot-count lval basr)]
+        (if (= sc 0)
```
keep wishing that clojure had a `nonzero?` predicate!

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+          cnt
+          (recur (+ cnt sc) (dec lind)))))))
+
+(defn- order-swap-parity [basl basr]
```
we should talk about this given yesterday's feeling that hestenic should remain a standalone(able) module. how does that assertion square with the important inclination to reuse code?

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (if (even? (order-swap-count basl basr)) +1 -1))
+
+(defn- sort-on-bases [bldds]
+  (sort (fn [bla blb] (compare (basis bla) (basis blb))) bldds))
```
wunderschoen! (did not formerly work, by some vague circumstance of... something, but now does)

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+(defn- same-rung? [this otha]
+  (instance? (class this) otha))
+
+(defn- different-rung? [this otha]
```
have removed `same-rung?`, since it doesn't get used -- thus eliminating the impropriety of defining `different-rung?` like this by eliminating its possibility.

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+;;
+
+(defprotocol IGradeling
+  (bladoids [this])
```
have an incipient bifurcated-semantics protocol use schema brewing... to be jointly chewed next we're in real time...

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+
+(defn- gradeling-absorb-bladoid [grdl bl]
+  (let [gra (grade grdl)]
+    (if (< gra 0)
```
ibid.

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (let [gra (grade grdl)]
+    (if (< gra 0)
+      (Gradeling. (grade bl) (list bl))
+      (if (not= gra (grade bl))
```
that would've forced the error clause below the healthy clause, no?

or were you suggesting deploying `(not (same-grade? ...`? i'd be game, but we've already 'calculated' the Gradeling's grade (as `gra`, in the surrounding `let`) and use it in a bunch of places. argument is that the typographic recurrence of `gra` throughout the function is a useful cue...

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+    (if (< gra 0)
+      (Gradeling. (grade bl) (list bl))
+      (if (not= gra (grade bl))
+        (throw (Exception. (str "can't absorb grade" (grade bl)
```
tasty! but, as with the comment... somewhere above, what do we want to in cases like this given the desire to keep _hestenic_ a standalone module?

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+        (if (bladoid-with-basis grdl (basis bl))
+          (Gradeling. gra (map (fn [blel]
+                                 (if (same-basis? bl blel)
+                                   ;; (sum bl blel)
```
think what's shaking out is a practice in which there's a cabal of backstage functions for calculating arithmetic stuff (including intermediate results) that we should keep closed under cabal membership (in the set-theoretic sense)... which means we shouldn't e.g. use `sum` in these `defn-` jobbies. know that there's an explicit transgression of this upstairs in `collapse`...

vanderoops replied:
think i may not follow... tawk?

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+                     "]")))
+
+(defn mv [gradels]
+  (MV. (sort-on-grades gradels)))
```
the comment's _arbitrary_ should probably have read _unconstrained_ or something.

the constituent Gradelings do want to be sorted (or fired through a dictionary, as you've recommended). idea of this constructor is that you can dump a pile of Gradelings in without pre-sorting them. though if the code goes this far, it should probably then also run the Gradeling-rung analogue of `collapse` on the resulting sorted collection...

vanderoops replied:
mathematically it'd work if it were jumbled. probably work algorithmically (given the current codepile) too. so strictly speaking sortedness is a convenience, but, hey, why not? certainly you don't want a random ordering for human readability...

if in a dictionary, i'm assuming that 'give me all the constituent Gradelings in order' undertakes the sort work again each time?

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+
+;;
+;; Vect is, of course, a vector (meaning -- yes! -- a real geometric,
+;; mathematical vector, for gods' sake; not an 'array', per lexical cretinism).
```
let's for sure discuss the idea of what constitutes probity in a codebase!

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+(deftype MV [grdlgs]
+  IMV
+  (gradelings [_] grdlgs)
+  (gradeling-of-grade [this gra]
```
mkay. is the recommendation to do with speed/efficiency? or idiomaticity? or?

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+;; simplification (and complexification) of species
+;;
+
+(extend-type Bladoid
```
indeed! for the moment i'm finding the rhythmic return of the code to the leftmost column for each type's implementation block to be visually useful...

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+;;
+;; all things Bladoid, don't you know. well, not all. in any event, one of these
+;; is a weighted (scaled) basis element of the GA. does not for the moment
+;; presuppose any particular dimension. also does not afford configurable
```
need to make some decisions around the two big pieces that want to be configurable: (1) dimensions, and (2) metric.

in most GA systems for compiled languages, these two choices get specified at compile time and then necessarily persist through the execution of the resulting program. that pretty much means you can't e.g. have a dimension-2 and -3 algebra running side by side in one program.

this here clojure certainly affords the opportunity for less ossified specification of dimension and metric. how exactly we want that stuff to work and feel's a good topic.

(noting here for time's posterior that at present the system is implicitly infinite-dimensional...)

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+;; metrics...
+;;
+
+(defprotocol IBladoid
```
going to wrap up all the other topics raised in your kind review and then return to this one (and the reader-invertible textual representation bits)...

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  (same-basis? [this otha]))
+
+
+(deftype Bladoid [cf bss]
```
a personal proscription is in play here. more on that via a different medium.

have adopted the following scheme: all fields' symbols' names will be the obvious ascii, prepended by `f_`.

so in this case we now have
```
(deftype Bladoid [f_coef f_basis]
  IBladoid
  (coef [_] f_coef)
  (basis [_] f_basis)
  ...
```

vanderoops replied:
... as to the three-letter naming schema's crypticicity -- one may note that way way down at the bottom of the source code there are now explicitly semantic expansions for all the IHestenic protocol's denizens. e.g.:

`(def grade-involution gri)`

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+    (let [gr (count bss)
+          flormp (/ (* gr (- gr 1)) 2)]
+      (if (even? flormp) +1 -1)))
+  (same-basis? [this otha]
```
please?

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,785 @@
+  ([blds gra]
+   (if (empty? blds)
+     blds
+     (loop [head (first blds)
```
apparently 'exclamation mark' is principally british in use, and 'exclamation point' american? not sure why i've really only ever heard the former...

anyway, have adopted one of those for this function's naming.

jackrusher replied:
> apparently 'exclamation mark' is principally british in use, and 'exclamation point' american? not sure why i've really only ever heard the former...

... but in our transnational culture, we know it as 'bang'.

Commit:
so much; so many: embetterments from sicumtils pr#347's review ...

... for which: thanks abundant to sritchie! many improvements stylistic;
 a few representational. more and then more to come!

Commit:
experiments with ouroborable representations for Bladoid and Vect...

... meaning that their printed forms can now be recognized by the
reader and hoovered back in. Congruent ouput versions too for Gradeling
and MV, though as yet unclear whether there's value in reader-specialized
input for these two.

Commit:
left and right contraction operations for ye olde GA.

Commit:
what's that you say? a configurable metric for GA operations? splendid!

Commit:
pseudoscalars, reversion operator, dualization; hestenes inner product...

... as well as supporting characters like the 'scalar?' predicate.

Commit:
grade involution, o my liege.

Commit:
and how could we've forgotten the ol' scalar product? well, here 'tis.

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,1084 @@
+(defmacro with-metric
+  [tric & formses]
+  `(binding [*hestenic-metric* ~tric]
+     ~(conj formses 'do)))
```
excellent; thanks! (and: done.)

vanderoops commented on src/sicmutils/hestenic.cljc:
```diff
@@ -0,0 +1,1084 @@
+       (if (not= gra (grade head))
+         (throw (Exception. (str "incompatible grade: "
+                                 head " but should be " gra)))
+         (if (not (empty? body))
```
prithee: wherefore?

Commit:
account for metric in calculating Bladoid and Vect inverses, and ...

additional convenience funq for accumulating metric factors. Because
we... care.

huahaiy commented:
This PR appears to be stalled. What's the holdup? Looking forward to using this to learn GA.

vanderoops commented:
Good day! And thanks for your interest in this GA implementation.

The 'hestenic' chunk of code in question needs a bunch of additional work before it would be appropriate to merge with the scimutils bloodline. And while my ardor for that project remains undimmed, I won't have time for a little while.

It's worth mentioning, though, that the design intent is to provide a set of GA tools that interoperates with the sicmutils system-of-systems but that also functions in a self-contained, standalone manner. And so... if your intent was to play around with GA in the context of sicmutils, then the news is not great short term (though eventually good);  but if you wanted to experiment with 'hestenic' by itself, I could certainly provide access to it in an isolated repo.

					john


> On Aug 14, 2022, at 10:59 AM, Huahai Yang ***@***.***> wrote:
> 
> 
> This PR appears to be stalled. What's the holdup? Looking forward to using this to learn GA.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
>

sritchie commented:
@huahaiy this was a boutique exploration into GA, as @vanderoops notes; if you're interested in learning GA with sicmutils, I encourage you to go for it!

What sort of materials are you looking for? An implementation of the types, sure, but beyond that, is there a book you're following along with, or something like that? Or are you looking for essays etc that go through the concepts, in addition to code?

huahaiy commented:
@vanderoops @sritchie Thanks for responding. I am not familiar with sicmutils at this moment, and certainly would be willing to dive in more, particularly if it has a good GA implementation, as that was my main interest. 

I have a few elementary books on GA, and they all mention the use of some software for some parts of the study. For example, Alan Macdonald's Linear and Geometric Algebra uses a GA module in SymPy; Miroslav Josipovic's Geometric Multiplication of Vectors uses Cl3 in Mathematica; Dorst et al.'s Geomtric algrebra for computer science discusses implementation of GA and has their own C++ implementation. 

I would ideally like an implementation of GA in Clojure to play with, hence my interest in this PR.

Commit:
make mult-inverse operator pervasive; add monograde? and invertible? mthds.

Commit:
ditch protocol for grade operations; fold those instead into IHestenic ...

... and, indeed, add both 'grade' and 'grades' methods. Also, rename
the grade selection (projection) operator from 'grade-part' to 'grp',
to underscore its fundamentalness and parity with other operators. But
then also alias the old prolix name back to the new ensveltened one.

Commit:
preferentially accordion gradlings to vects when of grade unity ...

... and also fix a shockingly naive-and-incomplete implementation of
dot product for Vect.

Commit:
explicitly require clojure.set, and move gradeling-as-vect downward, ...

... in order to support standalone loading (and by 'support' we mean
'enable at all without causing an error' (and by 'we' we mean 'i' (and
by 'mean' we mean 'stony-hearted'))).
